homeassistant/components/fortigate/__init__.py

```diff
@@ -0,0 +1,72 @@
+"""Fortigate integration."""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.const import CONF_HOST, CONF_USERNAME, CONF_PASSWORD, \
+    CONF_DEVICES, EVENT_HOMEASSISTANT_STOP
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.discovery import async_load_platform
+
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'fortigate'
+
+DATA_FGT = DOMAIN
+
+ATTR_NAME = 'name'
+DEFAULT_NAME = 'Default State'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_DEVICES, default=[]):
+            vol.All(cv.ensure_list, [cv.string]),
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+
+async def async_setup(hass, config):
+    """Start the fortigate component."""
+    conf = config.get(DOMAIN)
+
+    if conf is not None:
+
+        host = conf.get(CONF_HOST)
+        user = conf.get(CONF_USERNAME)
+        apikey = conf.get(CONF_PASSWORD)
+        devices = conf.get(CONF_DEVICES)
+
+        await async_setup_fortigate(hass, config, host, user, apikey, devices)
```

Comment:
I assume that there will be a traceback if the credentials are wrong. The setup should fails if the integration can't recover .

Reply:
If the login fails, the exception FGTConnectionError is raised at line 54

Reply:
We should `return False` from `setup` function if credentials are wrong since it's a non recoverable error. The user will need to change the config and restart.

Reply:
So I would need to merge the async_setup_fortigate and the async_setup, as the check of the credentials against the Fortigate is done at async_setup_fortigate ?

Reply:
We can return False from `async_setup_fortigate` if login fails and True at the end if everything is good. Then we can check the return value here in `setup`.

Reply:
fixed this, thanks